### PR TITLE
ac-34 product exports

### PR DESCRIPTION
**Pull request by ExtraGravity**

### CODE REVIEW

File: src/routers/productExports.ts [hunk 1]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
 const router = createRouter();
 
+/**
+ * @api {get} /productexports
+ * @description Returns all product exports and all their associated attributes.
+ *              Guaranteed fields are:
+ *                - productTransactionUuid
+ *                - productType
+ *                - datetime (UTC)
+ *                - toPersonUuid
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: use apiparam tag pls :)

File: src/routers/productExports.ts [hunk 2]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
+ *                - fromPersonUuid
+ *                - amountOfProduct
+ *                - productUnits
+ *                - costPerUnit
+ *                - currency
+ *              Parameters have no effect on this request.
+ *              Associated attributes can be checked via /products/
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: example of associate attribute for product exports? also i think its a different end point for product exports?

File: src/routers/productExports.ts [hunk 3]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
+ * @apiName GetProductExports
+ *
+ * @apiError (401) Unauthorized - Must be admin
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: 403

File: src/routers/productExports.ts [hunk 4]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
+ *
+ * @apiSuccess (200) {String} Successfully retrieved all product exports
+ * @apiSuccessExample Success-Response:
+  [
+    {
+      "productTransactionUuid": "9f11efb5-d5b4-4853-aa88-ee10c2940c9f",
+      "datetime": "2018-01-23 04:05:06",
+      "toPersonUuid": "34e0574e-beea-498d-8bec-7f26d84ba761",
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: beea!

File: src/routers/productExports.ts [hunk 5]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
+      "fromPersonUuid": "95bb2019-d135-49a8-83a7-4bf8f54684a0",
+      "productType": "milk",
+      "amountOfProduct": 10.23,
+      "productUnits": "litres",
+      "costPerUnit": 22.56,
+      "currency": "UGX",
+      "density": "5"
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: maybe provide density units?

File: src/routers/productExports.ts [hunk 6]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
+      "currency": "USD"
+    }
+  ]
+ */
 router.get('/', async (req, res) => {
   res.status(StatusCode.OK).send('Successfully retrieved all product exports') ;
 }, authorized(UserType.ADMIN));
 
+/**
+ * @api {get} /productexports
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: should be post

File: src/routers/productExports.ts [hunk 7]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
+ * @description Creates a new product export entry.
+ *              Required fields are:
+ *                - datetime (UTC)
+ *                - productType
+ *                - toPersonUuid
+ *                - fromPersonUuid
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: api param please!

File: src/routers/productExports.ts [hunk 8]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
+ *                - productType
+ *                - amountOfProduct
+ *                - costPerUnit
+ *                - currency
+ *              Along with any associated attributes.
+ *              Associated attributes can be checked via /products.
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: example?

File: src/routers/productExports.ts [hunk 9]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
+ *              The product UUID is returned.
+ * @apiName CreateProductExport
+ *
+ * @apiError (400) The following fields are missing, ...
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: specify it's a badrequest?

File: src/routers/productExports.ts [hunk 10]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
+ * @apiError (400) The product type does not exist
+ * @apiError (401) Unauthorized - Must be admin
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: forbidden

File: src/routers/productExports.ts [hunk 11]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
   res.status(StatusCode.CREATED).send('Successfully created new product export');
 }, authorized(UserType.ADMIN));
 
+/**
+ * @api {put} /productexports
+ * @description Updates a product export entry.
+ *              Required fields are:
+ *                - productTransactionUuid
+ *                - productType
+ *                - datetime (UTC)
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: apiparam please!

File: src/routers/productExports.ts [hunk 12]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
+ *                - toPersonUuid
+ *                - fromPersonUuid
+ *                - amountOfProduct
+ *                - costPerUnit
+ *                - currency
+ *              Along with any associated attributes.
+ *              Associated attributes can be checked via /products.
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: wrong endpoint?

File: src/routers/productExports.ts (hunk 8 above)
Comment (beaesguerra) [reply src/routers/productExports.ts]: wrong endpoint?

File: src/routers/productExports.ts [hunk 13]
```diff
@@ -7,14 +7,110 @@ import { StatusCode } from '@/models/statusCodes';
...
+ * @apiName UpdateProductExport
+ *
+ * @apiError (400) The folllowing attributes do not exist
+ * @apiError (401) Unauthorized - Must be admin
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: 403

File: src/routers/productExports.ts (hunk 12 above)
Comment (beaesguerra) [reply src/routers/productExports.ts]: how are we getting associated attributes anyways for product exports?

File: src/routers/products.ts [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+const router = createRouter();
+
+/**
+ * @api {get} /products
+ * @description Returns all product types, their unique associated attributes, and their units.
+ *              Parameters have no effect on this request.
+ * @apiName GetPeople
+ *
+ * @apiError (401) Unauthorized - Must be admin
```
Comment (beaesguerra) [inline src/routers/products.ts]: 403

File: src/routers/products.ts [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+ * @apiSuccessExample Success-Response:
+  milk: {
+      "units": "litres",
+      "attributes": [
+        "density"
+      ]
+  }
+ */
+router.get('/', async (req, res) => {
+  res.status(StatusCode.OK).send('Successfully retrieved all product exports') ;
```
Comment (beaesguerra) [inline src/routers/products.ts]: should be products type not exports

**Review by beaesguerra — CHANGES_REQUESTED** [review]

**Comment by ExtraGravity** [comment]
We don't need this yet so I'll fix it when we do.

File: src/routers/productExports.ts (hunk 5 above)
Comment (beaesguerra) [reply src/routers/productExports.ts]: do we need density units?

File: src/routers/productExports.ts [hunk 16]
```diff
@@ -6,18 +6,107 @@ import { StatusCode } from '@/models/statusCodes';
...
+    {
+      "productTransactionUuid": "3321269e-9b8b-432f-a668-b65c206235b0",
+      "datetime": "2018-01-23 04:05:20",
+      "toPersonUuid": "34e0574e-beea-498d-8bec-7f26d84ba761",
+      "fromPersonUuid": "95bb2019-d135-49a8-83a7-4bf8f54684a0",
+      "productType": "corn",
+      "amountOfProduct": 5.2,
+      "productUnits": "kilograms",
+      "costPerUnit": 2.2,
+      "currency": "USD"
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: do you think we should be doing export products by type?

File: src/routers/productExports.ts [hunk 17]
```diff
@@ -6,18 +6,107 @@ import { StatusCode } from '@/models/statusCodes';
...
+    }
+  ]
+ */
 router.get('/', async (req, res) => {
-  // TODO
   res.status(StatusCode.OK).send('Successfully retrieved all product exports') ;
 }, authorized(UserType.ADMIN));
 
+/**
+ * @api {post} /productexports
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: hehe i still think it would be nicer if it was /exports/products or products/exports or something so that its not multiple words. but that's just me!

File: src/routers/productExports.ts (hunk 7 above)
Comment (beaesguerra) [reply src/routers/productExports.ts]: api param :)

File: src/routers/productExports.ts [hunk 18]
```diff
@@ -6,18 +6,107 @@ import { StatusCode } from '@/models/statusCodes';
...
   res.status(StatusCode.CREATED).send('Successfully created new product export');
 }, authorized(UserType.ADMIN));
 
+/**
+ * @api {put} /productexports/:uuid
+ * @apiName UpdateProductExport
+ * @apiGroup ProductExports
+ * @apiVersion 0.0.1
+ * @apiDescription Updates a product export entry.
+ *              Possible fields are:
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: api params?

**Review by beaesguerra — COMMENTED** [review]
Re-reviewing. Some of my older comments havent been addressed still I think

File: src/routers/productExports.ts (hunk 5 above)
Comment (ExtraGravity) [reply src/routers/productExports.ts]: No, this is an extra attribute, so anything goes.

File: src/routers/productExports.ts (hunk 17 above)
Comment (ExtraGravity) [reply src/routers/productExports.ts]: It does look nicer, but I don't think it makes as much sense. Since product exports is it's own entity.

File: src/routers/productExports.ts (hunk 16 above)
Comment (ExtraGravity) [reply src/routers/productExports.ts]: Nah, not for now.

File: src/routers/productExports.ts (hunk 18 above)
Comment (ExtraGravity) [reply src/routers/productExports.ts]: Actually all the fields were wrong, must've been high.

**Comment by ExtraGravity** [comment]
Implementation is included in PR. But no testing with auth, couldn't get it to work even on other APIs.

File: dbscripts/db-init.sql [hunk 19]
```diff
@@ -208,8 +208,10 @@ CREATE TABLE ProductPayments (
 -- ProductExports
 
 CREATE TABLE ProductExports (
+	productExportUuid UUID PRIMARY KEY NOT NULL DEFAULT uuid_generate_v4(),
```
Comment (beaesguerra) [inline dbscripts/db-init.sql:211]: do we still need the uuid_generate_v4 if the client is sending us the uuid? 

don't have to do it as part of this change but i think since this has recently changed, we should be making sure in all of the requests that they provide the uuid (I don't think I have been checking for this in the past)

File: dbscripts/db-init.sql [hunk 20]
```diff
@@ -208,8 +208,10 @@ CREATE TABLE ProductPayments (
...
 	productTypeId SERIAL REFERENCES ProductTypes(productTypeId) NOT NULL,
 	amountOfProduct REAL NOT NULL,
 	transportId VARCHAR(255),
```
Comment (beaesguerra) [inline dbscripts/db-init.sql]: should this be not null?

File: src/database/ProductExports/index.ts [hunk 21]
```diff
@@ -0,0 +1,83 @@
+import dbConnection, { tableNames, execute } from '../connection';
+
+const prodExportTable = () => dbConnection()(tableNames.PRODUCT_EXPORTS);
+
+/**
+ * Represent a product export after retrieval from the database
+ */
+export interface ProdExportDb {
```
Comment (beaesguerra) [inline src/database/ProductExports/index.ts:8]: i dont think this should have "db" after because productname isn't in the product export table?

File: src/database/ProductExports/index.ts [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+  productexportuuid: string;
+  transportid: string;
+  datetime: string;
+  producttypeid: number;
+  amountofproduct: number;
+  lastmodified: string;
+}
+
+const builders = {
+  /** Get all product transactions of a certain type */
```
Comment (beaesguerra) [inline src/database/ProductExports/index.ts]: update comment

File: src/database/ProductExports/index.ts [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+  getSingleProdExport(productexportuuid: string) {
+    return prodExportTable().select('*')
```
Comment (beaesguerra) [inline src/database/ProductExports/index.ts]: does this need to return the product name? or is the id enough?

Comment (beaesguerra) [reply src/database/ProductExports/index.ts]: we should probably be consistent since the get all product exports api returns the productname too

File: src/database/ProductExports/index.ts [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+
+  /** Get all product exports */
+  getProdExports() {
+    return prodExportTable().select('*')
+    .join(tableNames.PRODUCT_TYPES,
+      tableNames.PRODUCT_TYPES + '.producttypeid',
+      tableNames.PRODUCT_EXPORTS + '.producttypeid');
+  },
+
+  /** Inserts a new produt export row in the database */
```
Comment (beaesguerra) [inline src/database/ProductExports/index.ts]: typo in product

File: dbscripts/db-init.sql [hunk 25]
```diff
@@ -208,8 +208,10 @@ CREATE TABLE ProductPayments (
 -- ProductExports
 
 CREATE TABLE ProductExports (
```
Comment (beaesguerra) [inline dbscripts/db-init.sql:210]: don't we need to track who exported the product? ie. which trader is adding it?

File: src/database/ProductExports/index.ts [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+
+/** Get all product exports of a certain type */
+export async function getProdExports(): Promise<ProdExportDb[]> {
+  const exports = await execute<ProdExportDb[]>(builders.getProdExports());
+  return exports;
+}
+
+/** creates a new product export in the database, returns the new UUID */
```
Comment (beaesguerra) [inline src/database/ProductExports/index.ts:80]: can we use proper js documentation? it usually auto generates it so that it requires you to specify what the params represent

Comment (beaesguerra) [reply src/database/ProductExports/index.ts:80]: and if there's a return value, it also lets you describe what its supposed to return

File: src/routers/index.ts [hunk 27]
```diff
@@ -13,5 +13,6 @@ router.use('/people', people);
 router.use('/products', products);
 router.use('/transactions', transactions);
 router.use('/productExports', productExports);
+router.use('/products', products);
```
Comment (beaesguerra) [inline src/routers/index.ts]: this is already in line 13

File: src/database/ProductExports/index.ts [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+export async function insertProdExport(req: ProdExportDbInsertReq): Promise<string> {
+  const newUuid = await execute<any>(builders.insertProdExport(req));
```
Comment (beaesguerra) [inline src/database/ProductExports/index.ts:82]: doesn't the server need to generate the last modified date?

File: src/database/ProductExports/index.ts [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+  return newUuid[0];
+}
+
+/** Update a single column for a single product export */
+export async function updateProdExportField(uuid: string, field: string, value: string|number) {
+  return await execute<any>(builders.updateProdExportField(uuid, field, value));
```
Comment (beaesguerra) [inline src/database/ProductExports/index.ts:88]: doesn't the server need to generate the last modified date?

File: src/routers/productExports.ts [hunk 30]
```diff
@@ -1,24 +1,171 @@
...
+}
+
+/**
+ * Represents a product exoprt udpate request in the API
+ */
+export interface ProdExportUpdateReq {
+  uuid: string;
+
+  transportId?: string;
+  datetime?: string;
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: we probably shouldn't let them update the datetime for when it was first created?

File: src/routers/productExports.ts [hunk 31]
```diff
@@ -1,24 +1,171 @@
...
+  datetime: string;
+  productType: string;
+  amountOfProduct: number;
+  lastModified: string;
+}
+
+/**
+ * Represents a product export creation request in the API
+ */
+export interface ProdExportReq {
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: can you make this ProdExportCreationReq so it's clear and consistent with the other interfaces too?

File: src/routers/productExports.ts [hunk 32]
```diff
@@ -1,24 +1,171 @@
...
+    "amountOfProduct": 5.2
+  }
+*/
 router.put('/:uuid', async (req, res) => {
-  // TODO
-  res.status(StatusCode.OK).send('Successfully updated product export');
-}, authorized(UserType.ADMIN));
+  const updateReq: ProdExportUpdateReq = {
+    uuid: req.params.uuid,
+
```
Comment (beaesguerra) [inline src/routers/productExports.ts:166]: extra space

File: src/routers/productExports.ts [hunk 33]
```diff
@@ -1,24 +1,171 @@
...
+    transportId: req.body.transportId,
+    datetime: req.body.datetime,
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: they shouldn't be able to change datetime? but maybe the client should be generating the last modified timestamp?

File: src/routers/productExports.ts [hunk 34]
```diff
@@ -1,24 +1,171 @@
...
+    amountOfProduct: req.body.amountOfProduct,
+  };
+
+  try {
+    await ProdExportsService.updateProdExportInDb(updateReq);
+  } catch (e) {
+    res.status(StatusCode.INTERNAL_SERVER_ERROR).send();
+  } finally {
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: should this be in a finally?  won't this still execute after an error is caught? how come this block of code can't be put in the try?

File: src/services/ProductExports.ts [hunk 35]
```diff
@@ -0,0 +1,134 @@
...
+import logger from '@/utilities/modules/logger';
+
+/** message when an unhandled error is thrown */
+export const unhandledErrorMsg = 'unhandled error';
+
+/**
+ * Connector for the API and database to get a single product export
+ */
+export async function getProdExportFromDb(uuid: string): Promise<api.ProdExport> {
+  logger.info('getting single transaction', uuid);
```
Comment (beaesguerra) [inline src/services/ProductExports.ts]: did we want to keep this logging message?

File: src/services/ProductExports.ts [hunk 36]
```diff
@@ -0,0 +1,134 @@
...
+
+
+/**
+ * Connector for the API and database to get all product exports
+ */
+export async function getProdExportsFromDb(): Promise<api.ProdExport[]> {
+  const dbProdExports: db.ProdExportDb[] = await db.getProdExports();
+  const results: api.ProdExport[] = [];
+
+  logger.info('results', results);
```
Comment (beaesguerra) [inline src/services/ProductExports.ts]: did we want to log this? it'll always log an empty array

File: src/services/ProductExports.ts [hunk 37]
```diff
@@ -0,0 +1,134 @@
...
+
+  dbProdExports.forEach(function (item: db.ProdExportDb) {
+    const newExport: api.ProdExport = {
+      uuid: item.productexportuuid,
+      transportId: item.transportid,
+      datetime: item.datetime,
+      productType: item.productname,
```
Comment (beaesguerra) [inline src/services/ProductExports.ts]: if we're already returning the productname, should we also return the product units? it should be in the item already anyways since it was joined with the products table

File: src/services/ProductExports.ts [hunk 38]
```diff
@@ -0,0 +1,134 @@
...
+ */
+export async function createProdExportInDb(apiReq: api.ProdExportReq): Promise<string> {
+  let productTypeId: number = -1;
+  const invalidFields: string[] = [];
+
+  // validate request
+  try {
+    productTypeId = await productTransactionsDb.getProductId(apiReq.productType);
+  } catch (e) {
+    throw new Error('Product type ' + apiReq.productType + ' not supported');
```
Comment (beaesguerra) [inline src/services/ProductExports.ts:66]: don't we also need to throw this error if producttypeid is still -1?

File: src/services/ProductExports.ts [hunk 39]
```diff
@@ -0,0 +1,134 @@
...
+    invalidFields.push('uuid');
+  }
+  if (!apiReq.transportId) {
+    invalidFields.push('transportId');
+  }
+  if (!apiReq.datetime) {
+    invalidFields.push('datetime');
+  }
+  if (!apiReq.productType) {
+    invalidFields.push('productType');
```
Comment (beaesguerra) [inline src/services/ProductExports.ts:82]: would it be an invalid field if the product type isn't valid? should we be treating that case like how we are treating these cases?

File: src/services/ProductExports.ts [hunk 40]
```diff
@@ -0,0 +1,134 @@
...
+  }
+  if (!apiReq.amountOfProduct) {
+    invalidFields.push('amountOfProduct');
+  }
+
```
Comment (beaesguerra) [inline src/services/ProductExports.ts:87]: if we know all the invalid fields by this point, shouldnt we throw an error now instead of trying to insert a request?

File: src/services/ProductExports.ts [hunk 41]
```diff
@@ -0,0 +1,134 @@
...
+  return newUuid;
+}
+
+/**
+ * Connector for the API and database to update a product export
+ * Throws error with message "unhandled error" for unhandled errors
+ */
+export async function updateProdExportInDb(apiReq: api.ProdExportUpdateReq) {
+  try {
+    if (apiReq.transportId) {
```
Comment (beaesguerra) [inline src/services/ProductExports.ts]: there's no way of doing an update of multiple fields in one query?

File: src/database/ProductExports/index.ts (hunk 28 above)
Comment (beaesguerra) [reply src/database/ProductExports/index.ts:82]: ignore this, i saw it later on in the service file

File: src/services/ProductExports.ts [hunk 42]
```diff
@@ -0,0 +1,134 @@
...
+      await db.updateProdExportField(apiReq.uuid, 'transportid', apiReq.transportId);
+    }
+    if (apiReq.datetime) {
```
Comment (beaesguerra) [inline src/services/ProductExports.ts]: we probabbly shouldn't allow this! :o

**Review by beaesguerra — CHANGES_REQUESTED** [review]

File: src/routers/productExports.ts (hunk 34 above)
Comment (ExtraGravity) [reply src/routers/productExports.ts]: Because the trader wants the actual correct info on error.

File: src/services/ProductExports.ts (hunk 38 above)
Comment (ExtraGravity) [reply src/services/ProductExports.ts:66]: I don't understand, the only time we need to throw this error is if we hit an error when we're getting the producttypeid.

File: dbscripts/db-init.sql (hunk 19 above)
Comment (ExtraGravity) [reply dbscripts/db-init.sql:211]: Yea we shouldn't. I copy pasted most of this code from the ProductTransactions page which I hacked together as fast as I can.

File: src/database/ProductExports/index.ts (hunk 21 above)
Comment (ExtraGravity) [reply src/database/ProductExports/index.ts:8]: I think it's more like, what will the product exports database interface is gonna give. Not exactly what's in the database. Because this can't just return an ID, it's gotta give a product name somehow.

File: src/database/ProductExports/index.ts (hunk 23 above)
Comment (ExtraGravity) [reply src/database/ProductExports/index.ts]: Good catch

File: dbscripts/db-init.sql (hunk 25 above)
Comment (ExtraGravity) [reply dbscripts/db-init.sql:210]: This wasn't what was originally agreed to. Is this something we definitely want?

File: src/database/ProductExports/index.ts (hunk 26 above)
Comment (ExtraGravity) [reply src/database/ProductExports/index.ts:80]: This is alot of documentation to update haha.

File: src/database/ProductExports/index.ts (hunk 29 above)
Comment (ExtraGravity) [reply src/database/ProductExports/index.ts:88]: Yup, missed it.

File: src/routers/productExports.ts (hunk 30 above)
Comment (ExtraGravity) [reply src/routers/productExports.ts]: I wanted to give the clients more freedom. There's alot of things that I think shouldn't be changeable but seems like people wanted anything to be changed, including money transactions.

File: src/services/ProductExports.ts (hunk 37 above)
Comment (ExtraGravity) [reply src/services/ProductExports.ts]: Sure

File: src/services/ProductExports.ts (hunk 40 above)
Comment (ExtraGravity) [reply src/services/ProductExports.ts:87]: The error thrwoing is being done before the actual insert request is done. But I created the insert request struct before throwing the error, I'll switch them around.

File: src/services/ProductExports.ts (hunk 41 above)
Comment (ExtraGravity) [reply src/services/ProductExports.ts]: Lazy :) 
I copied the code I hacked together. 

Let's optimize it later, we'll create a new issue for it and that way, muz can get some easy exposure to the backend that he was looking for.

File: dbscripts/db-init.sql (hunk 25 above)
Comment (beaesguerra) [reply dbscripts/db-init.sql:210]: recorderUuid

File: src/routers/productExports.ts (hunk 34 above)
Comment (beaesguerra) [reply src/routers/productExports.ts]: pls fix

File: src/database/ProductExports/index.ts (hunk 21 above)
Comment (beaesguerra) [reply src/database/ProductExports/index.ts:8]: pls fix

File: src/routers/productExports.ts (hunk 34 above)
Comment (beaesguerra) [reply src/routers/productExports.ts]: pls fix

File: src/database/ProductExports/index.ts (hunk 21 above)
Comment (ExtraGravity) [reply src/database/ProductExports/index.ts:8]: I don't remember what there was to fix.

File: src/database/ProductExports/index.ts (hunk 29 above)
Comment (ExtraGravity) [reply src/database/ProductExports/index.ts:88]: Oh it's done in the service actually.

File: src/routers/productExports.ts (hunk 32 above)
Comment (ExtraGravity) [reply src/routers/productExports.ts:166]: I realized this was intentional to separate mandatory and non mandatory fields.

File: src/middleware/index.ts [hunk 43]
```diff
@@ -2,16 +2,14 @@ import * as compression from 'compression';
...
-import unauthorizedMiddleware from '@/middleware/unauthorized';
 
 /**
  * Applies middleware to express app using old `configure` pattern
  */
 export default function middleware() {
   this.use(createNetworkLogger());
-  this.use(jwtMiddleware());
-  this.use(unauthorizedMiddleware());
+  // this.use(jwtMiddleware());
```
Comment (beaesguerra) [inline src/middleware/index.ts]: put these back

File: src/routers/productExports.ts [hunk 44]
```diff
@@ -33,9 +36,7 @@ export interface ProdExportReq {
  */
 export interface ProdExportUpdateReq {
   uuid: string;
-
   transportId?: string;
-  datetime?: string;
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: i thought we will let them change datetime? and shouldnt we also let them change the recorderuuid too?

File: src/routers/productExports.ts [hunk 45]
```diff
@@ -131,8 +138,6 @@ router.post('/', async (req, res) => {
  *
  * @apiParam {String} uuid Specify the UUID of the product export to update.
  * @apiParam [String] transportId Identifier for mode of transport, max 255 characters.
- * @apiParam [String] datetime The time the export occured.
- * @apiParam [String] productType The type of product.
```
Comment (beaesguerra) [inline src/routers/productExports.ts]: why was this removed?

File: src/routers/productExports.ts (hunk 32 above)
Comment (beaesguerra) [reply src/routers/productExports.ts:166]: you could add a comment instead?

File: src/services/ProductExports.ts [hunk 46]
```diff
@@ -119,9 +124,6 @@ export async function updateProdExportInDb(apiReq: api.ProdExportUpdateReq) {
     if (apiReq.transportId) {
       await db.updateProdExportField(apiReq.uuid, 'transportid', apiReq.transportId);
     }
-    if (apiReq.datetime) {
```
Comment (beaesguerra) [inline src/services/ProductExports.ts]: i thought we were letting them change datetime? and then should we let them change recorder uuid?

**Review by beaesguerra — CHANGES_REQUESTED** [review]

File: src/routers/productExports.ts (hunk 45 above)
Comment (ExtraGravity) [reply src/routers/productExports.ts]: They can't change the producttype

File: src/routers/productExports.ts (hunk 44 above)
Comment (ExtraGravity) [reply src/routers/productExports.ts]: I forgot :P

**Review by beaesguerra — APPROVED** [review]
